heralding/capabilities/mysql.py

```diff
@@ -0,0 +1,189 @@
+            zeros+salt_2+self.AUTH_PLUGIN
+
+        return packet
+
+    def auth_switch_request(self, seq_no):
+        # Auth Switch Request Packet:
+
+        # 0x00 0x03    Payload length
+        # 0x03 0x01    Sequence No
+        # 0x04 0x01    Auth switch request (0xFE)
+        # 0x05 0x16    Auth method name (variable length, NUL-terminated)
+        # 0x1B 0x15    Auth method data (20 bytes of random salt + \0)
+
+        # Incase any client using other auth methods, request client to switch to 'mysql_native_password'
+
+        packet_length = 3+1+1+len(self.AUTH_PLUGIN)+20+1
+        payload_len = MySQL.convert4To3Byte(packet_length-4)
+        seq_no = bytes([seq_no])
+        auth_switch_req = b'\xFE'
+        auth_data = os.urandom(20)+b'\x00'
+
+        packet = payload_len+seq_no+auth_switch_req+self.AUTH_PLUGIN+auth_data
+
+        return packet
+
+    def auth_failed(self, seq_no, user, server, using_password):
+        # Auth Failed Packet(Error 1045):
+
+        # 0x00 0x03    Payload length
+        # 0x03 0x01    Sequence #
+        # 0x04 0x01    Error packet (0xFF)
+        # 0x05 0x02    Error code (1045, 0x0415)
+        # 0x07 0x06    SQL State (#28000)
+        # 0x0D  ??     Error message
+
+        error_msg = bytes("Access denied for user '{}'@'{}' (using password: {})".format(
+            user, server, using_password), 'ascii')
+        full_length = 3+1+1+2+6+len(error_msg)  # taking out null
+
+        payload_len = MySQL.convert4To3Byte(full_length-4)
+        seq_no = bytes([seq_no])
+        error_packet = b'\xFF'  # Error Packet ID
+        error_code = b'\x15\x04'  # Error code 1045 (0x0415)
+        sql_state = bytes("#28000", 'ascii')
+
+        packet = payload_len+seq_no+error_packet+error_code+sql_state+error_msg
+        return packet
+
+    async def execute_capability(self, reader, writer, session):
+        try:
+            await self._handle_session(reader, writer, session)
+        except struct.error as exc:
+            logger.debug('MySQL connection error: %s', exc)
+            session.end_session()
+
+    async def _handle_session(self, reader, writer, session):
+        # HandshakeResponse41 Packet:
+
+        # 0x00 0x03    Payload length
+        # 0x03 0x01    Sequence No
+        # 0x04 0x04    Client capability flags
+        # 0x08 0x04    Max packet size
+        # 0x0C 0x01    Client character set
+        # 0x0D 0x17    Unused(zeros)
+        # 0x24 ??      User name (NUL-terminated)
+        # XX   0x01    Password length
+        # XX+1 ??      Password
+        # YY   ??      Schema (NUL-terminated)
+        # ZZ   ??      Client authentication plugin (NUL-terminated)
+        # RR   ??      Client connection attibutes
+
+        address = writer.get_extra_info('peername')[0]
+        writer.write(self.server_greeting())
+        data = await reader.read(2048)
+
+        caps = int.from_bytes(data[0x04:0x08], byteorder='little')
```

Comment:
You need to keep track of the length of data. Here we are using indexing (0x04), but are we sure that data even holds that many elements?